src/pdn/src/shape.cpp

```diff
@@ -0,0 +1,627 @@
+  }
+
+  // merge all to get most restrictive obstruction box
+  obs_.mergeInit();
+  obs_.merge(spacingdb_rect);
+  obs_.merge(spacing_table_rect);
+  obs_.merge(eol_rect);
+}
+
+const std::string Shape::getDisplayText() const
+{
+  const std::string seperator = ":";
+  std::string text;
+
+  text += net_->getName() + seperator;
+  text += layer_->getName() + seperator;
+  if (grid_component_ != nullptr) {
+    text += GridComponent::typeToString(grid_component_->type()) + seperator;
+    text += grid_component_->getGrid()->getName();
+  } else {
+    text += "none";
+  }
+
+  return text;
+}
+
+bool Shape::isRemovable() const
+{
+  if (!isModifiable()) {
+    return false;
+  }
+
+  if (getNumberOfConnections() < 2) {
+    // floating shape with one or zero connections
+    return true;
+  }
+
+  return false;
+}
+
+bool Shape::isModifiable() const
+{
+  return true;
+}
+
+const std::string Shape::getReportText() const
+{
+  return fmt::format("{} on {}",
+                     getRectText(rect_, layer_->getTech()->getLefUnits()),
+                     layer_->getName());
+}
+
+const std::string Shape::getRectText(const odb::Rect& rect,
+                                     double dbu_to_micron)
+{
+  return fmt::format("({:.4f}, {:.4f}) - ({:.4f}, {:.4f})",
+                     rect.xMin() / dbu_to_micron,
+                     rect.yMin() / dbu_to_micron,
+                     rect.xMax() / dbu_to_micron,
+                     rect.yMax() / dbu_to_micron);
+}
+
+Shape* Shape::extendTo(const odb::Rect& rect,
+                       const ShapeTree& obstructions) const
+{
+  std::unique_ptr<Shape> new_shape(copy());
+
+  if (isHorizontal()) {
+    new_shape->rect_.set_xlo(std::min(rect_.xMin(), rect.xMin()));
+    new_shape->rect_.set_xhi(std::max(rect_.xMax(), rect.xMax()));
+  } else if (isVertical()) {
+    new_shape->rect_.set_ylo(std::min(rect_.yMin(), rect.yMin()));
+    new_shape->rect_.set_yhi(std::max(rect_.yMax(), rect.yMax()));
+  } else {
+    return nullptr;
+  }
+
+  if (rect_ == new_shape->rect_) {
+    // shape did not change
+    return nullptr;
+  }
+
+  if (obstructions.qbegin(bgi::intersects(new_shape->getRectBox())
+                          && bgi::satisfies([this](const auto& other) {
+                               // ignore violations that results from itself
+                               return other.second.get() != this;
+                             }))
+      != obstructions.qend()) {
+    // extension not possible
+    return nullptr;
+  }
+
+  return new_shape.release();
+}
+
+/////////
+
+FollowPinShape::FollowPinShape(odb::dbTechLayer* layer,
+                               odb::dbNet* net,
+                               const odb::Rect& rect)
+    : Shape(layer, net, rect, odb::dbWireShapeType::FOLLOWPIN)
+{
+}
+
+Shape* FollowPinShape::copy() const
+{
+  auto* shape = new FollowPinShape(getLayer(), getNet(), getRect());
+  shape->generateObstruction();
+  shape->rows_ = rows_;
+  return shape;
```

Comment:
Could this be replaced with a call to the copy-ctor, e.g. new FollowPinShape(*this); ?  Are there any fields you are not copying?

Similarly with Shape::Copy

Reply:
I'm using the copy function to make a copy of either Shape or FollowPinShape (overwritten), so I'm not sure if I can just use the copy-ctor since I sometimes need Shape and other times it's child class FollowPinShape.

Reply:
I see copy is virtual so never mind.